src/implementations/svd.jl

```diff
@@ -203,7 +242,40 @@ function svd_trunc!(A::AbstractMatrix, USVᴴ, alg::TruncatedAlgorithm)
     return truncate!(svd_trunc!, USVᴴ′, alg.trunc)
 end
 
-### GPU logic
+# Diagonal logic
+# --------------
+function svd_full!(A::AbstractMatrix, USVᴴ, alg::DiagonalAlgorithm)
+    check_input(svd_full!, A, USVᴴ, alg)
+    Ad = diagview(A)
+    U, S, Vᴴ = USVᴴ
+    p = sortperm(Ad; by=abs, rev=true)
+    zero!(U)
+    zero!(Vᴴ)
+    T = eltype(U)
+    U[CartesianIndex.(enumerate(p))] .= Ref(one(T))
+    Vᴴ[CartesianIndex.(reverse.(enumerate(p)))] .= sign_safe.(view(Ad, p))
```

Comment:
I don't know if we want to support `U` and `Vᴴ` that don't have efficient linear indexing. With that restriction in mind, one solution that would have slightly fewer allocations could be:
```julia
p = sortperm(Ad; by=abs, rev=true)
ip = p .+ (0:n-1) .* n
p .= (1:n) .+ (p .- 1) .* n
U[p] .= one(T)
Vᴴ[ip] .= sign_safe.(view(Ad, p))
```

Not sure if that is fully correct; I haven't checked/tested this, but I guess the idea is clear.

Reply:
I'm slightly confused about this, doesn't my current implementation have no allocations since the entire broadcasting expression should fuse, while your suggested implementation has one additional vector allocated? I agree that the linear indexing might be more efficient, but it's hard to say without testing